### DISCUSSION TITLE
Filtering Data Provider on selected rows from previous screen

### DISCUSSION THREAD

**Discussion by oldwired** [Budibase/budibase#8759] [Help] [answered]
I have 2 (MariaDB) tables in my Application. "Events" and "Data". Now I want users to be able to select 1 to many Events via checkboxes on the first screen and have them display that selection (enriched with some of the data from the data table) after a button press on a different screen.

So for example if the "Event table" on screen1 looks like this:
<img width="108" alt="image" src="https://user-images.githubusercontent.com/9700404/203112182-6f724237-e521-4987-8592-2d6b684129f7.png">

I want them to be able to select some rows, press the button and have them shown on screen2 like this:
```
Event ABC
*Additional Data read from data table*
 
Event ASD
*Additional Data read from data table*
```

I tried the following:

1. Enable "Allow row selection" on screen1
2. Add a Button to screen 1 within the dataprovider with 2 actions:
  a) "Update State" Set Value Key:"sel" Value: {{ eventgrid.Selected rows }}
  b) "Navigate To" /screen2
3. On screen2 add a data provider, a repeater within and to try it text that shows the event "{{ eventrepeater.events.Ereignis }}"
4. Try to define a filter for the data provider on screen2. But on what should I filter? The {{ State.sel }} is   `%5B6192497041519406%5D,%5B6192497041333655%5D,%5B6192497041317846%5D` if I select 3 rows.

**Comment by melohagan** [comment, marked as answer]
Hey @oldwired 

Is this what you were hoping to achieve?
![selected_rows](https://user-images.githubusercontent.com/101575380/203342013-72a1a209-835e-4e38-8c0f-2f2900af6cc8.gif)


### Home screen
Add a Data provider with table that allows for row selection:
<img width="260" alt="Screenshot 2022-11-22 at 14 40 14" src="https://user-images.githubusercontent.com/101575380/203342273-26bcada7-7a23-40c0-ad14-17fef0c5bd57.png">
Add a button underneath the table.

**Define actions** for the button click as follows:
<img width="920" alt="Screenshot 2022-11-22 at 14 41 16" src="https://user-images.githubusercontent.com/101575380/203342479-8a094524-f29f-48c0-878d-d93f0443974e.png">

### Detail screen
Define the route of the detail screen to include the selected rows as a variable:
<img width="262" alt="Screenshot 2022-11-22 at 14 42 05" src="https://user-images.githubusercontent.com/101575380/203342672-2de2016e-1d51-435a-a136-a246e6b1ce96.png">

Add a card block for the same table as in the home screen. You don't have to use a card block, but it's convenient for this example.
<img width="262" alt="Screenshot 2022-11-22 at 14 43 41" src="https://user-images.githubusercontent.com/101575380/203343011-97325f94-b818-4eb3-aed1-68e5a60b3454.png">

Supply the bindings for the title, subtitle and description.
<img width="262" alt="Screenshot 2022-11-22 at 14 44 45" src="https://user-images.githubusercontent.com/101575380/203343295-b5f1f452-65d9-41b0-a4d8-437ceaf5dc85.png">
E.g. `{{ New Cards block.Sales.Order ID }}`

Finally add a filter to the card block as follows:
<img width="926" alt="Screenshot 2022-11-22 at 14 45 52" src="https://user-images.githubusercontent.com/101575380/203343551-82b8bc09-fffa-42c6-9bf4-d467f1e946cd.png">

#### App export
[Filter on selected rows-export-1669127853715.tar.gz](https://github.com/Budibase/budibase/files/10068430/Filter.on.selected.rows-export-1669127853715.tar.gz)

**Reply by oldwired** [reply]
Yes, this looks like what I am trying. I must have missed the "_id". I can only try this out tomorrow, though. I will come back with feedback.
I would guess this is a common usecase for building reports with drilldowns of selected positions.

**Reply by melohagan** [reply]
There is also this link, which is more to do with relational data, but a similar thing: https://docs.budibase.com/docs/filter-by-relationships

**Reply by oldwired** [reply]
If I select "_id" I get a "Unknown column 'events._id' in 'where clause'" error message.
<img width="194" alt="image" src="https://user-images.githubusercontent.com/9700404/203543158-73755d1c-080c-440b-950b-814cddb6b65c.png">

But when I look at what lands in my URL it's "[6192497041860180],[6192497041835996],[6192497041727338]" 
<img width="232" alt="image" src="https://user-images.githubusercontent.com/9700404/203543483-e7611cf4-e62f-4824-b6a7-12f766f7b822.png">
and those numbers are values from my "PID" column. The column is my primary key in this table, so that makes sense. So I tried to change my filter to the PID column
<img width="517" alt="image" src="https://user-images.githubusercontent.com/9700404/203543246-aedc3438-aad6-4d91-ad96-31aa0eb3ba07.png">
But then I get a new error
<img width="190" alt="image" src="https://user-images.githubusercontent.com/9700404/203543559-652dd5ba-ba5c-48f3-a98e-0cf5585b9eef.png">
and the result is "No rows found"

**Reply by oldwired** [reply]
The table is created as follows. (MyISAM because it's just for reporting, I need no referential integrity, etc but fast bulk inserts). The order of the columns is strange (for historical) reasons, but that should not matter.
```SQL
CREATE TABLE `protocols`.`events` (
	`tname` VARCHAR(13) NOT NULL COLLATE 'utf8mb4_general_ci',
	`tidx` INT(11) NOT NULL,
	`oid` INT(11) NULL DEFAULT NULL,
	`Objektname` VARCHAR(80) NULL DEFAULT NULL COLLATE 'utf8mb4_general_ci',
	`PID` BIGINT(20) NOT NULL,
	`Zeit` DATETIME NULL DEFAULT NULL,
	`Ereignis` VARCHAR(40) NULL DEFAULT NULL COLLATE 'utf8mb4_general_ci',
	`Massnahme` VARCHAR(40) NULL DEFAULT NULL COLLATE 'utf8mb4_general_ci',
	`Meldung` VARCHAR(120) NULL DEFAULT NULL COLLATE 'utf8mb4_general_ci',
	PRIMARY KEY (`PID`) USING BTREE,
	INDEX `Ereignis` (`Ereignis`) USING BTREE,
	INDEX `oid_idx` (`oid`) USING BTREE,
	INDEX `Zeit` (`Zeit`) USING BTREE
)
 COLLATE 'utf8mb4_general_ci' ENGINE=MyISAM ROW_FORMAT=Dynamic;
```

**Reply by oldwired** [reply]
I think the square brackets are causing trouble, because if I try it without them it works:
<img width="520" alt="image" src="https://user-images.githubusercontent.com/9700404/203548626-b4568fa0-d39b-4bbf-81a9-681d55356cc6.png">
Is there any way to get rid of the square brackets?

Edit: I'm currently testing `{{ replace URL.selectedrows '[' '' }}`, but it looks like I can't nest them.

Edit2: I just `replace` the left (though it's URI encoded), save it to a state variable, then the right one and use the state variable on the next screen.

Edit3: with the stripped brackets I can use it to filter on my PID and it works. :)

**Reply by melohagan** [reply]
Instead of using **_id** you may be able to use the actual name of the primary key instead. May be related to this bug: https://github.com/Budibase/budibase/issues/8334

**Reply by oldwired** [reply]
Yes, I think it's this same bug.
I already switched to filtering on my primary key (luckily it's a single int64 column, I had a composite key in previous iterations but changed that, because I could not get CSV export working #7855 

But to be able to use the "Selected Rows" I had to strip ti of the (URI encoded) square brackets [ ] first.
